### DISCUSSION TITLE
Wrapping `fast-check` as a DSL for topic-specific tests

### DISCUSSION THREAD

**Discussion by ornamentist** [dubzzz/fast-check#3714] [Q&A] [answered]
I'm really enjoying `fast-check` for property based testing in JavaScript and TypeScript.  

I'd like to take things further by wrapping property tests in a topic-specific testing DSL in TypeScript:

```
type Swans = ...

theorem("all swans are white", swans(100), (s: Swan) => {
  ...
});
```

Where `theorem` would wrap say a Mocha test and `fast-check` properties via function overloads:

```
import * as Mocha from "mocha";
import * as Check from "fast-check";

function theorem<T>(a1: Check.Arbitrary<T>, c: (x: T) => boolean | void): void;
function theorem<T, U>(a1: Check.Arbitrary<T>, a2: Check.Arbitrary<U>, c: (x: T, y: U) => boolean | void): void;
// more function overloads

function theorem<T, U = T,...>(?): void {
  Mocha.it("foo", () => Check.assert(Check.property(?)));
}
```

I'm not sure how to type the `theorem` function body. Should I be using the user-facing `fast-check` types based on `<T extends [unknown, ...unknown[]]>` or the same approach that test runners use in wrapping `fast-check`?

Any advice much appreciated.

**Comment by dubzzz** [comment, marked as answer]
You can probably have a look to the typing declared for `test.prop` in `@fast-check/{vitest,jest}` (in this repository). They seem pretty close to your needs.

**Reply by ornamentist** [reply]
Thanks -- `fast-check` is a seriously useful project.

**Reply by ornamentist** [reply]
This looks useful:

```
// ...

/**
 * prop has just been declared for typing reasons, ideally TestProp should be enough
 * and should be used to replace `{ prop: typeof prop }` by `{ prop: TestProp<???> }`
 */
declare const prop: <Ts, TsParameters extends Ts = Ts>(
  arbitraries: Ts extends [any] | any[] ? ArbitraryTuple<Ts> : ArbitraryRecord<Ts>,
  params?: FcParameters<TsParameters>
) => (
  testName: string,
  prop: Ts extends [any] | any[] ? Prop<Ts> : PropRecord<Ts>,
  timeout?: number | undefined
) => void;
```
